### PR TITLE
[games] Implement in-game saves

### PR DESCRIPTION
**Pull request by fetzerch**
Implement in-game (save ram) saves.

## Description
This implements in-game saves where the frontend accesses the game's memory and saves/loads it to/from disk. This is mostly used in emulators for SRAM (battery backed up ram on cartridges) or memory cards.

For reference, these are the other ways how saving is implemented in the Game API:
- Savestates (https://github.com/xbmc/xbmc/pull/11034)
- The game saves it directly (already implemented and used for example by ScummVM)

Differences to save states:
- Works only for supported games (e.g. emulated games with SRAM support)
- Often works emulator independent (and can be used to start a game with one emulator and continue with another)
- Visible in-game (e.g. in-game save game selection menus)
- Closer to how saving in the good old days worked where one could only save at certain points where the game allowed it.

Previously PRed to the retroplayer banch in https://github.com/garbear/xbmc/pull/73
*API-Change*: Merging has to be synced with https://github.com/kodi-game/game.libretro/pull/10.

## Motivation and Context
Save games.

## How Has This Been Tested?
Best tested with 2048 or SNES/GBA emulators. 

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/games/addons/GameClient.h [hunk 1]
```diff
@@ -161,6 +162,9 @@ class CGameClient : public ADDON::CAddonDll
   std::unique_ptr<IGameClientPlayback> m_playback; // Interface to control playback
   GAME_REGION           m_region;              // Region of the loaded game
 
+  // In-game saves
+  std::unique_ptr<CGameClientInGameSaves> m_inGameSaves;
```
Comment (notspiff) [inline xbmc/games/addons/GameClient.h]: why a pointer? it's unconditionally allocated.

File: xbmc/games/addons/GameClientInGameSaves.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  m_gameClient(addon),
+  m_dllStruct(dllStruct)
+{
+  assert(m_gameClient != nullptr);
+  assert(m_dllStruct != nullptr);
+}
+
+void CGameClientInGameSaves::Load()
+{
+    Load(GAME_MEMORY_SAVE_RAM);
```
Comment (notspiff) [inline xbmc/games/addons/GameClientInGameSaves.cpp]: indent

File: xbmc/games/addons/GameClientInGameSaves.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    Load(GAME_MEMORY_RTC);
+}
+
+void CGameClientInGameSaves::Save()
+{
+    Save(GAME_MEMORY_SAVE_RAM);
```
Comment (notspiff) [inline xbmc/games/addons/GameClientInGameSaves.cpp]: indent

File: xbmc/games/addons/GameClientInGameSaves.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  }
+  catch (...)
+  {
+    CLog::Log(LOGERROR, "GAME: %s: Exception caught in GetMemory()", m_gameClient->ID().c_str());
+  }
+
+  if (size > 0)
+  {
+    const std::string path = GetPath(memoryType);
+
```
Comment (notspiff) [inline xbmc/games/addons/GameClientInGameSaves.cpp]: why the spacefest?

File: xbmc/games/addons/GameClientInGameSaves.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+struct GameClient;
+
+namespace GAME
+{
+  class CGameClient;
+
+  /*!
+   * \brief This class implements in-game saves.
+   *
+   * Some games do not implement state persistence on their own, but rely on the frontend for saving their current
```
Comment (notspiff) [inline xbmc/games/addons/GameClientInGameSaves.h]: i'd do \details here. not exactly a brief.

Comment (fetzerch) [inline xbmc/games/addons/GameClientInGameSaves.h]: brief ends when starting a new paragraph (see https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmddetails and http://www.stack.nl/~dimitri/doxygen/manual/docblocks.html#cppblock). but why not be more specific

File: xbmc/games/addons/GameClient.h (hunk 1 above)
Comment (fetzerch) [inline xbmc/games/addons/GameClient.h]: I tried that at first, but CGameClientInGameSaves needs a valid pointer to the dll struct which is initialized only in CGameClient::Initialize so I have to delay the initialization.

File: xbmc/games/addons/GameClientInGameSaves.h (hunk 5 above)
Comment (akva2) [inline xbmc/games/addons/GameClientInGameSaves.h]: yeah i know but i like being specific as you say (every time i encountered it i had to look up in the manual - has started to stick now though ;))

File: xbmc/games/addons/GameClientInGameSaves.cpp (hunk 4 above)
Comment (fetzerch) [inline xbmc/games/addons/GameClientInGameSaves.cpp]: sorry that one i don't get. any more indentation errors, or did i mess up with the size somehow?

Comment (akva2) [inline xbmc/games/addons/GameClientInGameSaves.cpp]: 3 newlines seems excessive.

Comment (fetzerch) [inline xbmc/games/addons/GameClientInGameSaves.cpp]: That is very true ;) sry i looked at GetPath() in Load()

**Comment by akva2** [comment]
https://imgflip.com/i/1h66wu

**Comment by akva2** [comment]
(and it looks all good to me now).

**Comment by fetzerch** [comment]
Thanks!

Does it make sense to save periodically (maybe every minute)? We don't control cores and they might crash before we're able to stop a game. Guess that could easily be implemented with a CTimer in CGameClient and could even be reused later for savestates. If it makes sense this can however come in a separate PR.

**Comment by akva2** [comment]
maybe, but then we'd definitely need to ping-pong on the sram file, or you may end up dumping corrupted data.

**Comment by garbear** [comment]
I would like to abstract the various software and hardware details of managing one's saved data.

I think the best interface would be a "Saved game" concept associated with a single game/file. The saved game would include the player's current progress, including awards/achievements (scraped from a ROM achievements service) and a history of saved data. The saved data would include all forms of the game's storage at a point in time, and we could eventually allow more fine-grained memory management like loading/saving/importing/exporting SRAM saves (per-file) or savestates (per-emulator).

File: xbmc/games/addons/GameClientInGameSaves.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+#include "utils/URIUtils.h"
+#include "utils/log.h"
+
+#include <assert.h>
+
+using namespace GAME;
+
+#define INGAME_SAVES_DIRECTORY          "InGameSaves"
+#define INGAME_SAVES_EXTENSION_SAVE_RAM ".sav"
+#define INGAME_SAVES_EXTENSION_RTC      ".rtc"
```
Comment (thebrid) [inline xbmc/games/addons/GameClientInGameSaves.cpp]: Would you consider replacing these with typed constants?

**Comment by garbear** [comment]
@fetzerch is this ready to go in?

File: xbmc/games/addons/GameClient.cpp [hunk 7]
```diff
@@ -276,6 +277,9 @@ bool CGameClient::OpenFile(const CFileItem& file, IGameAudioCallback* audio, IGa
   if (!InitializeGameplay(file.GetPath(), audio, video))
     return false;
 
+  m_inGameSaves.reset(new CGameClientInGameSaves(this, &m_struct));
```
Comment (garbear) [inline xbmc/games/addons/GameClient.cpp]: Move this to the constructor. I was just bit by a class that got the order of closing the resource wrong, so it's best to enforce the condition that it's always allocated.

Comment (fetzerch) [inline xbmc/games/addons/GameClient.cpp]: See the comment above from from spiff, I would have done it if it was possible, but m_struct is only valid after CGameClient::Initialize.

**Comment by fetzerch** [comment]
From my side it's ready - but I thought you don't want it in its current from (without the abstraction).
Imho it would be nice to have save games working in nightly-builds and we can refactor it to a more fancy solution afterwards, but then at least the code doing the actual saving/loading is there.
Another advantage would be that we don't have to sync the game.libretro API change any more. It's always something you have to remember when rebasing the retroplayer branches.

Therefore: Your button, (needs a merge of kodi-game/game.libretro#10 as well).

**Review by garbear — APPROVED** [review]

**Comment by garbear** [comment]
jenkins build this please

**Pull request merged**